File: examples/kamon-1.x/get-started-sbt/src/main/java/kamon/examples/java/TracesSubscriptions.java
package kamon.examples.java;


import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.actor.UntypedActor;
import kamon.Kamon;
import kamon.trace.SegmentInfo;
import kamon.trace.TraceInfo;
import kamon.trace.Tracer;

import java.util.List;
import static scala.collection.JavaConversions.seqAsJavaList;

public class TracesSubscriptions {
  public static void main(String[] args) throws InterruptedException {
    Kamon.start();

    final ActorSystem mySystem = ActorSystem.create("my-system");
    final ActorRef subscriber = mySystem.actorOf(Props.create(TracePrinter.class), "subscriber");

    // tag:subscribe-to-traces:start
    Kamon.tracer().subscribe(subscriber);
    // tag:subscribe-to-traces:end


    // tag:create-some-traces-and-segments:start
    Tracer.withNewContext("trace-with-segments", true, () -> {
      Tracer.currentContext().withNewSegment("quick-segment", "code", "kamon", () -> {
        try{ Thread.sleep(100); } catch (Exception e) { e.printStackTrace(); }
        return 0;
      });

      Tracer.currentContext().withNewSegment("slow-segment", "code", "kamon", () -> {
        try{ Thread.sleep(3000); } catch (Exception e) { e.printStackTrace(); }
        return 0;
      });

      return "done";
    });
    // tag:create-some-traces-and-segments:end



    // Wait a bit for the snapshot to be flushed.
    Thread.sleep(15000);

    Kamon.shutdown();
  }


  // tag:subscriber:start
  public static class TracePrinter extends UntypedActor {

    @Override
    public void onReceive(Object message) throws Exception {
      if(message instanceof TraceInfo) {
        final TraceInfo traceInfo = (TraceInfo) message;
        final List<SegmentInfo> segments = seqAsJavaList(traceInfo.segments());

        System.out.println("#################################################");
        System.out.println("Trace Name: " + traceInfo.name());
        System.out.println("Timestamp: " + traceInfo.timestamp());
        System.out.println("Elapsed Time: " + traceInfo.elapsedTime());
        System.out.println("Segments: ");

        segments.forEach(segmentInfo -> {
          System.out.println("    ------------------------------------------");
          System.out.println("    Name: " + segmentInfo.name());
          System.out.println("    Category: " + segmentInfo.category());
          System.out.println("    Library: " + segmentInfo.library());
          System.out.println("    Timestamp: " + segmentInfo.timestamp());
          System.out.println("    Elapsed Time: " + segmentInfo.elapsedTime());
        });

      } else unhandled(message);
    }
  }
  // tag:subscriber:end
}

File: examples/kamon-1.x/get-started-sbt/src/main/scala/kamon/examples/scala/TracesSubscriptions.scala
package kamon.examples.scala

import akka.actor.{ActorSystem, Props, Actor}
import kamon.Kamon
import kamon.metric.SubscriptionsDispatcher.TickMetricSnapshot
import kamon.metric.TraceMetrics
import kamon.metric.instrument.Histogram
import kamon.trace.{TraceInfo, Tracer}


object TracesSubscriptions extends App {
  Kamon.start()
  val mySystem = ActorSystem("my-system")
  val subscriber = mySystem.actorOf(Props[TracePrinter], "subscriber")

  // tag:subscribe-to-traces:start
  Kamon.tracer.subscribe(subscriber)
  // tag:subscribe-to-traces:end

  // tag:create-some-traces-and-segments:start
  Tracer.withNewContext("example-trace", autoFinish = true) {
    Tracer.currentContext.withNewSegment("quick-segment", "code", "kamon") {
      Thread.sleep(100)
    }

    Tracer.currentContext.withNewSegment("slow-segment", "code", "kamon") {
      Thread.sleep(3000)
    }
  }
  // tag:create-some-traces-and-segments:end

  // Wait a bit for the snapshot to be flushed.
  Thread.sleep(15000)
  Kamon.shutdown()
}

// tag:subscriber:start
class TracePrinter extends Actor {
  def receive = {
    case traceInfo: TraceInfo =>

    println("#################################################");
    println("Trace Name: " + traceInfo.name)
    println("Timestamp: " + traceInfo.timestamp)
    println("Elapsed Time: " + traceInfo.elapsedTime)
    println("Segments: ");

    traceInfo.segments.foreach { segmentInfo =>
      println("    ------------------------------------------");
      println("    Name: " + segmentInfo.name)
      println("    Category: " + segmentInfo.category)
      println("    Library: " + segmentInfo.library)
      println("    Timestamp: " + segmentInfo.timestamp)
      println("    Elapsed Time: " + segmentInfo.elapsedTime)
    }
  }
}
// tag:subscriber:end

/*
// tag:example-output:start
#################################################
Trace Name: example-trace
Timestamp: 1429311913406000000.nanos
Elapsed Time: 3103544112.nanos
Segments:
    ------------------------------------------
    Name: quick-segment
    Category: code
    Library: kamon
    Timestamp: 1429311913408589398.nanos
    Elapsed Time: 100184837.nanos
    ------------------------------------------
    Name: slow-segment
    Category: code
    Library: kamon
    Timestamp: 1429311913509367747.nanos
    Elapsed Time: 3000135844.nanos
// tag:example-output:end
*/